core/trigonometry.cpp

```diff
@@ -11,56 +11,170 @@
 
 namespace scipp::core {
 
-Variable to_rad(Variable var) {
-  const auto scale = makeVariable<double>(Dims(), Shape(),
-                                          units::Unit(units::rad / units::deg),
-                                          Values{pi<double> / 180.0});
-  return var * scale;
-}
+const auto deg_to_rad =
+    makeVariable<double>(Dims(), Shape(), units::Unit(units::rad / units::deg),
+                         Values{pi<double> / 180.0});
 
-Variable sin(const Variable &var) {
+Variable sin(const VariableConstProxy &var) {
   using std::sin;
-  const Variable &rad = var.unit() == units::deg ? to_rad(var) : var;
+  Variable out(var);
+  if (var.unit() == units::deg)
+    out *= deg_to_rad;
```

Comment:
This is forcing an extra copy now, so this should be slower than before?

Reply:
Yes. I'm struggling to find a clean way to handle this that does not end up with more code duplication than I'd like.

The thought did occur to me though that should this even do a unit conversion if we are following the principle of small isolated operations?

Reply:
I see your point about not doing a unit conversion. In this case, however, this is generic enough to be justified. No one would be surprise of `sin(90 deg)` works.

Regarding the copy, better (but not ideal):
1. Create `out` as done in your changes.
2. Forward to call with out arg. This avoids the second copy, and should thus be much closer to optimal performance.

We still end up *iterating* the array twice (copy, then apply), even if no copy is required, but iteration should be considerably cheaper than copies involving memory allocation.